### PR TITLE
My account plan limits overview

### PR DESCRIPTION
**Pull request by daisyzarzo**
relates to https://github.com/bespoken/dashboard/issues/848

### CODE REVIEW

**Comment by daisyzarzo** [comment]
image
image

File: src/components/MyAccount/Account.test.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ describe("Account component", function () {
                 endingCard={""}
                 expirationDate={""}
                 capturedLogs={"0/10M"}
```
Comment (jperata) [inline src/components/MyAccount/Account.test.tsx]: this number should come from the logs limit

Comment (daisyzarzo) [reply src/components/MyAccount/Account.test.tsx]: this is just a test, for this purpose it doesn't matter

File: src/services/subscription-plan.ts [hunk 2]
```diff
@@ -46,20 +46,18 @@ export const getPrice = async (plan: string) => {
     return "0";
 };
 
-export const checkUserPlanLimits = async (plan: string, stripeSubscribedPlanId: string,
-    sources: any[], userEmail: string) => {
-
+export const getLogsCapturedThisMounth = async (plan: string, stripeSubscribedPlanId: string,
```
Comment (jperata) [inline src/services/subscription-plan.ts]: getLogsCapturedThisMonth

File: src/services/subscription-plan.ts [hunk 3]
```diff
@@ -69,13 +67,39 @@ export const checkUserPlanLimits = async (plan: string, stripeSubscribedPlanId:
...
 
-        for (const iterator of sources) {
-            const helper = (await logService.getLogsBySource(iterator.source.secretKey
-                , startTime, endTime));
-            if (helper.data) {
-                totalLogsCounted += helper.data;
+        if (sources && sources.length > 0) {
+            if (sources[0].source) {
+                for (const iterator of sources) {
+                    const helper = (iterator.source.hasIntegrated && await logService.getLogsBySource(iterator.source.secretKey, startTime, endTime));
```
Comment (jperata) [inline src/services/subscription-plan.ts]: this is the same code as below, you can extra into a small function and pass iterator.source as a parameter or iterator depending on which one it is.

getTotalLogsCounted(source, startTime, endTime)

Comment (jperata) [inline src/services/subscription-plan.ts]: don't use helper, describe what this variable is

File: src/pages/MyAccountPage.tsx [hunk 4]
```diff
@@ -45,37 +52,63 @@ export class MyAccountPage extends React.Component<MyAccountPageProps, MyAccount
...
+            expirationDate = card.exp_month + "/" + helper;
+            cardType = card.brand;
+            cardLast4Numbers = card.last4;
         }
+
+        const planData = await payment.getPlansData();
+
+        helper = await payment.getLogsCapturedThisMounth(plan, userDetail.stripeSubscribedPlanId, this.props.sources) + "";
+        let helper2: String = planData[lowerCasePlan].benefits.logs + "";
+        capturedLogs = helper + "/" + helper2.substr(0, helper2.length - 6) + "M";
```
Comment (jperata) [inline src/pages/MyAccountPage.tsx]: this only works for numbers in the millions, extract to a small function that generate the string, if the number is bigger than a million you can do a int division to get how many millions there are and then append the "M", if it's less than a million you can put the whole number.

File: src/services/subscription-plan.ts (hunk 2 above)
Comment (daisyzarzo) [reply src/services/subscription-plan.ts]: XD

File: src/components/MyAccount/Account.test.tsx [hunk 5]
```diff
@@ -38,7 +38,7 @@ describe("Account component", function () {
...
```
[the remark this answers is not in the document]
Comment (jperata) [reply src/components/MyAccount/Account.test.tsx]: agreed with previous comment, this is just setup for the test.

File: src/components/MyAccount/Account.test.tsx [hunk 6]
```diff
@@ -51,7 +51,7 @@ describe("Account component", function () {
             <Account email={""} nextBillingDate={"nextBillingDate"} cardType={""} currentPlan={"Standard"}
                 endingCard={""} expirationDate={""}
                 capturedLogs={"0/10M"}
```
[the remark this answers is not in the document]
Comment (jperata) [reply src/components/MyAccount/Account.test.tsx]: agreed with previous comment, this is just setup for the test.

**Comment by daisyzarzo** [comment]
image

image

File: package.json [hunk 7]
```diff
@@ -135,4 +135,4 @@
     "stripe": "^6.0.0",
     "uuid": "3.0.1"
   }
-}
+}
```
Comment (jperata) [inline package.json]: no reason to change this one

File: src/pages/MyAccountPage.tsx [hunk 8]
```diff
@@ -35,6 +42,13 @@ function mapDispatchToProps(dispatch: Redux.Dispatch<any>) {
         },
     };
 }
+export const getNumberInEnglishFormat = (x: number) => {
+    const stringNumber = x + "";
+    if (stringNumber.length > 6) {
```
Comment (jperata) [inline src/pages/MyAccountPage.tsx]: no reason to compare strings, just do x > 999999 and remove the stringNumber variable

File: src/pages/MyAccountPage.tsx [hunk 9]
```diff
@@ -45,39 +59,66 @@ export class MyAccountPage extends React.Component<MyAccountPageProps, MyAccount
...
+            monitoredSources: "",
+            virtualDevicesAllowed: "",
         };
     }
     async componentDidMount() {
+        let expirationDate = "";
+        let cardLast4Numbers = "";
+        let nextBillingDate = "";
+        let cardType = "";
+        let capturedLogs = "";
```
Comment (jperata) [inline src/pages/MyAccountPage.tsx]: this is only used after the if, so there is no reason to declare it as a let here, declare as a const in the point you are using it.

File: src/pages/MyAccountPage.tsx [hunk 10]
```diff
@@ -45,39 +59,66 @@ export class MyAccountPage extends React.Component<MyAccountPageProps, MyAccount
...
+        let monitoredSources = "";
```
Comment (jperata) [inline src/pages/MyAccountPage.tsx]: same

File: src/pages/MyAccountPage.tsx [hunk 11]
```diff
@@ -45,39 +59,66 @@ export class MyAccountPage extends React.Component<MyAccountPageProps, MyAccount
...
+        let virtualDevicesAllowed = "";
```
Comment (jperata) [inline src/pages/MyAccountPage.tsx]: same

File: src/pages/MyAccountPage.tsx [hunk 12]
```diff
@@ -45,39 +59,66 @@ export class MyAccountPage extends React.Component<MyAccountPageProps, MyAccount
...
         const userDetail = await auth.currentUserDetails();
-        if (userDetail && userDetail.stripeCustomerObjId && userDetail.stripeSubscribedPlanId) {
+        let plan = userDetail.stripeSubscribedPlanName ? userDetail.stripeSubscribedPlanName : "Free";
```
Comment (jperata) [inline src/pages/MyAccountPage.tsx]: const

File: src/services/subscription-plan.ts [hunk 13]
```diff
@@ -69,13 +76,37 @@ export const checkUserPlanLimits = async (plan: string, stripeSubscribedPlanId:
...
+};
+
+export const checkUserPlanLimits = async (plan: string, stripeSubscribedPlanId: string,
+    sources: any[], userEmail: string) => {
+    let totalLogsCounted = 0;
+
+    try {
+
+        const allowedLog = this.getLogsLimitNumber(plan);
+        totalLogsCounted = await this.getLogsCapturedThisMounth(plan, stripeSubscribedPlanId, sources);
```
Comment (jperata) [inline src/services/subscription-plan.ts]: this.getLogsCapturedThisMonth

File: src/services/subscription-plan.ts [hunk 14]
```diff
@@ -69,13 +76,37 @@ export const checkUserPlanLimits = async (plan: string, stripeSubscribedPlanId:
...
 
-        for (const iterator of sources) {
-            const helper = (await logService.getLogsBySource(iterator.source.secretKey
-                , startTime, endTime));
-            if (helper.data) {
-                totalLogsCounted += helper.data;
+        if (sources && sources.length > 0) {
+            if (sources[0].source) {
+                for (const iterator of sources) {
+                    if (iterator.source.hasIntegrated) {
```
Comment (jperata) [inline src/services/subscription-plan.ts]: you can include this as part of the getLogsCountedInAPeriod function, if the element .hasIntegrated is false then you return 0

File: src/services/subscription-plan.ts (hunk 13 above)
Comment (jperata) [reply src/services/subscription-plan.ts]: This typo should have broken a test , ensure this is tested

File: src/services/subscription-plan.ts (hunk 14 above)
Comment (daisyzarzo) [reply src/services/subscription-plan.ts]: the function is count amount of logs in a period of time, it does not matter if it is integrated or not

putting the integrated flag, saves you from entering the function so it saves you from entering another function
and it's a for entering a function, it's better if the flag is outside, just like this

**Comment by daisyzarzo** [comment]
relates to https://github.com/bespoken/dashboard/issues/868

**Comment by daisyzarzo** [comment]
image

File: src/services/subscription-plan.ts [hunk 15]
```diff
@@ -37,6 +37,14 @@ export class Payment {
         return 100;
     }
 
+    public getNumberOfVirtualDevices = (sources: any[]) => {
+        let totalVirtualDevicesCounted: number = 0;
+        for (const iterator of sources) {
+            if (iterator.silentEchoToken) totalVirtualDevicesCounted++;
```
Comment (jperata) [inline src/services/subscription-plan.ts]: this logic is wrong, if the user have multiple validations enabled is going to be counted multiple times.

**Pull request merged**